src/logfile_values_experimental.py

```diff
@@ -0,0 +1,224 @@
+def yield_metric_value(line, metric, time_format):
+    fields = line.split()
+    fields[0] = convert_to_timestamp(fields[0], time_format)
+    if fields[0] > (metric.now - metric.get_timeshift()):
+        return fields[metric.column]
+    else:
+        return False
+
+
+def read_logfile_reverse(filename, time_format, metric, buf_size=8192):
+    """
+    A generator that returns the lines of a file in reverse order.
+    Stops to read when has met timestamp bigger then desired period 
+    
+    returns False if file was not read completely
+    """
+    with open(filename) as fh:
+        segment = None
+        offset = 0
+        fh.seek(0, os.SEEK_END)
+        file_size = remaining_size = fh.tell()
+        while remaining_size > 0:
+            offset = min(file_size, offset + buf_size)
+            fh.seek(file_size - offset)
+            buffer = fh.read(min(remaining_size, buf_size))
+            remaining_size -= buf_size
+            lines = buffer.splitlines()
+            # the first line of the buffer is probably not a complete line so
+            # we'll save it and append it to the last line of the next buffer
+            # we read
+            if segment is not None:
+                # if the previous chunk starts right from the beginning of line
+                # do not concact the segment to the last line of new chunk
+                # instead, yield the segment first
+                if buffer[-1] is not '\n':
+                    lines[-1] += segment
+                else:
+                    yield yield_metric_value(segment, metric, time_format)
+            segment = lines[0]
+            for index in range(len(lines) - 1, 0, -1):
+                if len(lines[index]):
+                    yield yield_metric_value(lines[index], metric, time_format)
+    # Don't yield None if the file was empty
+    if segment is not None:
+        yield yield_metric_value(segment, metric, time_format)
+
+
+def convert_to_timestamp(time_str, time_format):
+    try:
+        timestamp = int(
+            time.mktime(time.strptime(time_str.split("+")[0], time_format)))
+    except ValueError:
+        try:
+            timestamp = int(time_str)
+        except ValueError:
+            timestamp = int(
+                time.mktime(
+                    time.strptime('-'.join(time_str.split("-")[:-1]),
+                                  time_format)))
+        pass
+    return int(timestamp)
+
+
+def main():
+    args = parse_args()
+    if args.debug:
+        logging.getLogger().setLevel(logging.DEBUG)
+        logging.getLogger().addHandler(logging.StreamHandler())
+
+    if args.timezone:
+        os.environ['TZ'] = args.timezone
+        # time.tzset()
+
+    for metric in args.metric:
+        template = args.prefix + '.{} {} ' + str(metric.now)
+        file_was_readed = True
+        # Read from the end of file until the timestamp is satisfying conditions
+        data = read_logfile_reverse(args.file, args.time_format, metric)
```

Comment:
It is not nice to pass the `time_format` and `metric` to this function.  We should get the lines and do the analyses in here.

Reply:
then how i ll understand that file was or not read completely

Reply:
You can iterate the generator as long as you like.